sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala

```diff
@@ -288,29 +285,105 @@ class HadoopTableReader(
     }
   }
 
+  /**
+   * The entry of creating a RDD.
+   */
+  private def getRDD(
+    inputClassName: String,
+    localTableDesc: TableDesc,
+    inputPathStr: String): RDD[Writable] = {
+    if (isCreateNewHadoopRDD(inputClassName)) {
+      createNewHadoopRdd(
+        localTableDesc,
+        inputPathStr,
+        inputClassName)
+    } else {
+      createHadoopRdd(
+        localTableDesc,
+        inputPathStr,
+        inputClassName)
+    }
+  }
+
   /**
    * Creates a HadoopRDD based on the broadcasted HiveConf and other job properties that will be
    * applied locally on each slave.
    */
   private def createHadoopRdd(
     tableDesc: TableDesc,
     path: String,
-    inputFormatClass: Class[InputFormat[Writable, Writable]]): RDD[Writable] = {
+    inputClassName: String): RDD[Writable] = {
 
     val initializeJobConfFunc = HadoopTableReader.initializeLocalJobConfFunc(path, tableDesc) _
 
     val rdd = new HadoopRDD(
       sparkSession.sparkContext,
       _broadcastedHadoopConf.asInstanceOf[Broadcast[SerializableConfiguration]],
       Some(initializeJobConfFunc),
-      inputFormatClass,
+      getInputFormat(inputClassName),
       classOf[Writable],
       classOf[Writable],
       _minSplitsPerRDD)
 
     // Only take the value (skip the key) because Hive works only with values.
     rdd.map(_._2)
   }
+
+  /**
+   * Creates a HadoopRDD based on the broadcasted HiveConf and other job properties that will be
+   * applied locally on each slave.
+   */
+  private def createNewHadoopRdd(
+    tableDesc: TableDesc,
+    path: String,
+    inputClassName: String): RDD[Writable] = {
+
+    val initializeJobConfFunc = HadoopTableReader.initializeLocalJobConfFunc(path, tableDesc) _
+
+    val newJobConf = new JobConf(hadoopConf)
+    initializeJobConfFunc.apply(newJobConf)
+    val rdd = new NewHadoopRDD(
+      sparkSession.sparkContext,
+      getNewInputFormat(inputClassName),
+      classOf[Writable],
+      classOf[Writable],
+      newJobConf
+    )
+
+    // Only take the value (skip the key) because Hive works only with values.
+    rdd.map(_._2)
+  }
+
+  /**
+   * If `spark.sql.hive.fileInputFormat.enabled` is true, this function will optimize the input
+   * method while reading Hive tables.
+   * For old input format `org.apache.hadoop.mapred.InputFormat`.
+   */
+  private def getInputFormat(
```

Comment:
those three functions below has a couple of lines and they are only called once. I wouldn't make a separate function. The codes can be moved into the functions above with some comments.